### DISCUSSION TITLE
My engine is mssql+pyodbc, failed to commit data across libraries when using the before_delete event

### DISCUSSION THREAD

**Discussion by Qiii103** [sqlalchemy/sqlalchemy#6664] [Usage Questions]
```
event.listen(SomeClass, 'before_delete', before_delete)

def before_delete(mapper, connect, target):
	# mapper2 is another connection
	ldb = mapper2
	object_id = target._id
	object_code = target.object_code
	proj_code = target.project.proj_code
	operator = user_code
	resource = ldb.Resource.query().filter(ldb.Resource.ref_id == object_id, ldb.Resource.active).one_or_none()
	if resource:
		resource.delete(operator)
```

The 'target' and 'resource' in before_delete do not belong to the same database connection
And finally it will execute 'ldb.commit() mapper2.commit()', 
but, only when using the before_delete event resource is not deleted

**Comment by Qiii103** [comment]
sqlalchemy version=1.3.20

**Comment by zzzeek** [comment]
hey there -

you can't safely run a Session.delete() operation inside of the before_delete() event.   See the notes at https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_delete#sqlalchemy.orm.MapperEvents.before_delete .

to run session operations such as session.delete() inside an event, you should use events like:

https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.before_flush
https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.after_flush_postexec

**Reply by Qiii103** [reply]
OK, thank you!!
I wonder if this has anything to do with the database,because I found that MySQL can successfully commit the deletion

**Reply by zzzeek** [reply]
when you say "across libraries", im not even sure if your operation above is dealing with two separate sessions simultaneously, because I would need a lot more context.  if the operations above are on the same session, maybe the order of operations happens to be different on one environment vs. another and the "deletes" for Resource objects occur after your event, so they get covered in one environment and not the other (this is part of why it's not safe to use Session operations within the flush).    Finally MySQL will autocommit if you are using MyISAM tables but I suspect in this case this has more to do with the DELETE happening or not.   Turning on echo=True in your engine will illustrate the series of events if you want to look into why it's happening.

You can also keep using before_delete() if you just manually emit a DELETE for your Resource object using the given Connection, like `connection.execute(delete(Resource).where(Resource.id == <pk>))`.   You could just run that instead.

**Reply by Qiii103** [reply]
First, I can determine that the object triggering the deletion event with the Resource object belong to two different sessions.Also, the two tables are in two different libraries.
The problem is to submit Session B in the event of session A. the submission of Session B fails!
I try to use the detached_to_persistent event in session A, the submission of Session B also fails
